src/AutoRest.CSharp/Common/Input/MergeDerivedModelsVisitor.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace AutoRest.CSharp.Common.Input
+{
+    internal class MergeDerivedModelsVisitor : InputNamespaceVisitor
+    {
+        private readonly IReadOnlyDictionary<InputModelType, InputModelType> _modelsToReplace;
+        private readonly HashSet<InputModelType> _baseModelsToUpdate;
+
+        private MergeDerivedModelsVisitor(IReadOnlyDictionary<InputModelType, InputModelType> modelsToReplace)
+        {
+            _modelsToReplace = modelsToReplace;
+            _baseModelsToUpdate = modelsToReplace.Keys.Select(k => k.BaseModel!).Distinct().ToHashSet();
+        }
+
+        public static InputNamespace Visit(InputNamespace rootNamespace)
+        {
+            if (rootNamespace.Models.All(static m => m.BaseModel is not {} bm || bm.DerivedModels.Contains(m)))
+            {
+                return rootNamespace;
+            }
+
+            var modelsToReplace = rootNamespace.Models
```

Comment:
A model inherits a base model, but not in `baseModel.DerivedModels`. This looks like an issue in TSP compiler or our emitter. Shall we fix it in compiler or emitter?

Reply:
https://github.com/Azure/autorest.csharp/issues/3876 - we need to decide what to do with it. However, since correct model graph is critical for consolidation work, I'll keep the workaround until 3876 is resolved.